web_src/js/features/shared.js

```diff
@@ -0,0 +1,38 @@
+import $ from 'jquery';
+
+const {csrfToken} = window.config;
+
+export function initSettingVariables() {
+  $('.show-variable-edit-modal').on('click', function () {
+    const target = $(this).attr('data-modal');
+    const $modal = $(target);
+    // clear input/textarea value
+    $modal.find('input[name=name]').val('');
+    $modal.find('textarea[name=data]').val('');
+    // set dialog header
+    const $header = $modal.find('#variable-edit-header');
+    $header.text($(this).attr('data-modal-header'));
+
+    if ($(this).attr('data-is-new') === 'false') {
+      // edit variable dialog
+      const oldName = $(this).attr('data-old-name');
+      const oldValue = $(this).attr('data-old-value');
+      $modal.find('input[name=name]').val(oldName);
+      $modal.find('textarea[name=data]').val(oldValue);
+    }
+
+    const url = $(this).attr('data-base-action');
+    const commitButton = $modal.find('.actions > .ok.button');
+    $(commitButton).on('click', (e) => {
+      e.preventDefault();
+      $.post(url, {
+        _csrf: csrfToken,
+        name: $modal.find('input[name=name]').val(),
+        data: $modal.find('textarea[name=data]').val(),
+      }, (data) => {
+        if (data.redirect) window.location.href = data.redirect;
+        else window.location.reload();
+      });
+    });
+  });
+}
```

Comment:
Here's this function rewritten without jQuery, please try it:

```js
export function initSettingVariables() {
  const btn = document.querySelector('.show-variable-edit-modal');
  if (!btn) return;

  btn.addEventListener('click', () => {
    const modal = document.querySelector(btn.getAttribute('data-modal'));
    const nameInput = modal.querySelector('input[name="name"]');
    const dataInput = modal.querySelector('textarea[name="data"]');

    // clear input/textarea value
    nameInput.value = '';
    dataInput.value = '';

    // set dialog header
    const header = modal.querySelector("#variable-edit-header");
    header.textContent = btn.getAttribute('data-modal-header');
    
    // edit variable dialog
    if (btn.getAttribute('data-is-new') === 'false') {
      nameInput.value = btn.getAttribute('data-old-name');
      dataInput.value = btn.getAttribute('data-old-value');
    }

    const url = btn.getAttribute('data-base-action');
    modal.querySelector('.actions .ok.button').addEventListener('click', async (e) => {
      e.preventDefault();
      const res = await fetch(url, {
        method: "POST",
        headers: {
          'content-type': 'application/json',
        },
        body: JSON.stringify({
          _csrf: csrfToken,
          name: nameInput.value,
          data: dataInput.value,
        });
      });
      const data = await res.json();
      if (data.redirect) {
        window.location.href = data.redirect;
      } else {
        window.location.reload();
      }
    });
  });
}
```

Reply:
I do not agree here.

1. There are not only one button. `querySelector` is not right, `querySelectoAll` is complex, jQuery simplifies the code a lot
2. The `.ui.modal` should be shown be Fomantic UI (aka jQuery) in the end, it doesn't sense to drop jQuery here.
3. If every new modal dialog needs so many lines of new code, should there be a framework level support?

Reply:
The `    modal.querySelector('.ok.button').addEventListener('click', async (e) => {` is the improper way to handle a modal.

Ideally, it should be `$(.ui.modal).....{onApprove: ...}` , that's what a framework should be.

Reply:
> There are not only one button. querySelector is not right, querySelectoAll is complex, jQuery simplifies the code a lot

I like the explicit DOM APIs that require to specify whether one or more results are expected. It eliminates a certain class of bugs.

> The .ui.modal should be shown be Fomantic UI (aka jQuery) in the end, it doesn't sense to drop jQuery here.
> If every new modal dialog needs so many lines of new code, should there be a framework level support?

Yeah I guess a wrapper function would makes sense.